state/allcollections.go

```diff
@@ -319,7 +319,11 @@ func allCollections() CollectionSchema {
 
 		// These collections hold information associated with machines.
 		containerRefsC: {},
-		instanceDataC:  {},
+		instanceDataC: {
+			indexes: []mgo.Index{{
+				Key: []string{"model-uuid", "machineid"},
```

Comment:
I'd question if we should have another index around the `instanceid` to ensure [this](https://github.com/juju/juju/blob/develop/state/machine.go#L1349) can be fast as well

Reply:
That access pattern is only used once in a machine's lifetime by `SetProvisioned`. I don't think there's a great need for it.

Reply:
`count, err := coll.Find(bson.D{{"instanceid", id}}).Count()` can be use just the index making the query extremely fast. Even though it's used once, it's still a full collection scan and for large deployments might be worth it?

Reply:
I've added the `model-uuid` + `instanceid` index.